dvc/scm.py

```diff
@@ -123,3 +125,60 @@ def resolve_rev(scm: "Git", rev: str) -> str:
             if len(ref_infos) > 1:
                 raise RevError(f"ambiguous Git revision '{rev}'")
         raise RevError(str(exc))
+
+
+def iter_revs(
+    scm: "Git",
+    head_revs: Optional[List[str]] = None,
+    num: int = 1,
+    all_branches: bool = False,
+    all_tags: bool = False,
+    all_commits: bool = False,
+    all_experiments: bool = False,
+) -> Mapping[str, str]:
+    from dvc.repo.experiments.utils import fix_exp_head
+
+    if num < 1 and num != -1:
+        raise InvalidArgumentError(f"Invalid number of commits '{num}'")
+
+    if not any(
+        [head_revs, all_branches, all_tags, all_commits, all_experiments]
+    ):
+        return {}
+
+    head_revs = head_revs or []
+    revs = []
+    for rev in head_revs:
+        revs.append(rev)
+        n = 1
+        while True:
+            if num == n:
+                break
+            try:
+                head = fix_exp_head(scm, f"{rev}~{n}")
+                assert head
+                revs.append(resolve_rev(scm, head))
+            except RevError:
+                break
+            n += 1
+
+    if all_commits:
+        revs.extend(scm.list_all_commits())
+    else:
+        if all_branches:
+            revs.extend(scm.list_branches())
+
+        if all_tags:
+            revs.extend(scm.list_tags())
+
+    if all_experiments:
+        from dvc.repo.experiments.utils import exp_commits
+
+        revs.extend(exp_commits(scm))
+
+    rev_resolver = partial(resolve_rev, scm)
+    result = {
+        sha: ",".join(name)
+        for sha, name in group_by(rev_resolver, revs).items()
+    }
+    return result
```

Comment:
I think we should leave formatting for the names here up to the caller, and having access to the actual names themselves would make this a more useful API call in general. The `",".join(names)` is really brancher specific behavior and should happen in `brancher()` not `iter_revs`

This method could actually just return the `group_by` dict itself, and any caller that actually cares about the names can get `items()` as needed. (So the return type would be `Mapping[str, Iterable[str]]` or `Dict` instead of `Mapping`)

```suggestion
    return group_by(rev_resolver, revs)
```

This way in callers where we only care about the rev SHAs (like exp show), we can just use the SHAs like `for rev in iter_revs():`.